### ISSUE TITLE
`kfp-api` individual integration tests fail (potential race condition)

### ISSUE THREAD

**Issue by DnPlas** [canonical/kfp-operators#243]
`kfp-api` [individual integration tests](https://github.com/canonical/kfp-operators/blob/main/charms/kfp-api/tests/integration/test_charm.py) fail due to `kfp-api` going into error state with the following message:

```
unit-kfp-api-0: 07:29:37 ERROR unit.kfp-api/0.juju-log Failed to generate container configuration.
unit-kfp-api-0: 07:29:37 ERROR unit.kfp-api/0.juju-log Failed to handle <LeaderElectedEvent via KfpApiOperator/on/leader_elected[4]> with error: Please add required database relation: eg. relational-db
unit-kfp-api-0: 07:29:39 ERROR unit.kfp-api/0.juju-log Failed to generate container configuration.
unit-kfp-api-0: 07:29:39 ERROR unit.kfp-api/0.juju-log Failed to handle <PebbleReadyEvent via KfpApiOperator/on/ml_pipeline_api_server_pebble_ready[7]> with error: Please add required database relation: eg. relational-db
unit-kfp-api-0: 07:29:41 ERROR unit.kfp-api/0.juju-log Failed to generate container configuration.
unit-kfp-api-0: 07:29:41 ERROR unit.kfp-api/0.juju-log Failed to handle <ConfigChangedEvent via KfpApiOperator/on/config_changed[10]> with error: Please add required database relation: eg. relational-db
unit-kfp-api-0: 07:34:04 ERROR unit.kfp-api/0.juju-log Failed to generate container configuration.
unit-kfp-api-0: 07:34:04 ERROR unit.kfp-api/0.juju-log Failed to handle <UpdateStatusEvent via KfpApiOperator/on/update_status[16]> with error: Please add required database relation: eg. relational-db
```

For a full log go [here](https://github.com/canonical/kfp-operators/actions/runs/5505194002/jobs/10033719818#step:5:269). 

This happens intermittently in the GitHub CI when the charm is deployed and a `wait_for_idle()` call is made, but we cannot discard an issue on a regular deployment in production.

**Comment by DnPlas** [comment]
It seems like this issue and [this other one](https://github.com/canonical/kfp-operators/issues/243#issuecomment-1647645819) are caused by a series of factors, but mainly because of the units in the model not being active and idle when certain test cases run. The following describes these factors:

1) The reported issue is caused by the model not being active and idle, in fact the test case that fails is [`TestCharm.test_prometheus_grafana_integration`](https://github.com/canonical/kfp-operators/actions/runs/5505194002/job/14903739584#step:5:319), which times out because prometheus-k8s needs a significant amount of time to deploy. This scenario was already reported and fixed by canonical/kfp-operators#178, but then reverted by canonical/kfp-operators#197. Bringing back the big timeout prevents this error.

2) [In the case of this comment](https://github.com/canonical/kfp-operators/issues/243#issuecomment-1647645819), the logs show the following message `unit-kfp-api-0: 10:10:31 ERROR unit.kfp-api/0.juju-log mysql:1: Failed to handle <RelationJoinedEvent via KfpApiOperator/on/mysql_relation_joined[28]> with error: List of <ops.model.Relation object-storage:2> versions not found for apps: minio`, which after closer inspection is caused by the `minio` unit not being active. [This CI run](https://github.com/canonical/kfp-operators/actions/runs/5692291406/job/15429010748#step:4:134) shows exactly this situation and confirms that `minio` is in waiting status, and that happens because, even if we use `wait_for_idle()` because of juju/python-libjuju#900, the wait won't have much effect. This issue is now fixed in python-libjuju 2.9.44. After using such version in the testing environment, this problem went away.

3) Not reported in this issue, but sometimes the mysql-k8s charm will not deploy correctly because the [current memory constraints](https://github.com/canonical/kfp-operators/blob/main/charms/kfp-api/tests/integration/test_charm.py#L59). This can also cause the integration tests to fail when relating `kfp-api` to `mysql-k8s`, as the `kfp-api` charm depends entirely on a working DB for it to be active and idle.
The data platforms team has suggested to increase that constraint or to use `mysql-k8s` from `8.0/edge`, a version that has a `testing` option, which automatically constraints the charm so it can be used in machines with limited resources. Using this version of `mysql-k8s` with the testing option enabled has also improved the way these tests are running, and has prevented the integration tests from failing.

All of the above is being worked in #273

**Comment by DnPlas** [comment]
After debugging and reproducing this issue in the GH runners, it can be concluded that the error message in the description of this issue comes from any charm that uses SDI and whose conditions to handle interfaces are not met. For instance, `kfp-api` will throw similar message when calling [`_get_interface()`](https://github.com/canonical/kfp-operators/blob/main/charms/kfp-api/src/charm.py#L335) while the charm's dependencies are not ready (mysql-k8s, minio, etc.).

The readiness of a charm's dependencies is related to many things, but as mentioned in my [previous comment](https://github.com/canonical/kfp-operators/issues/243#issuecomment-1655840361), in this case we have a combination of factors where `minio` and `mysql-k8s` cannot be deployed correctly. In #273 I have experimented with different ways to prevent this, and so far it has worked.

Finally, the error message is just SDI saying it cannot handle the relation, but that is really not representative of the real issue, so if anyone runs into something similar, special attention must be taken to other charms in the model that may affect the charm throwing this error. We could definitely improve the error message to make it less misleading.

**Comment by DnPlas** [comment]
Should be fixed by #273
